src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java

```diff
@@ -215,13 +261,77 @@ else if (targetY <= levelDepth && workerY > levelDepth)
                 //First calculate way to miner building.
                 newProxy = getProxy(buildingPos, worker.getPosition(), BlockPosUtil.getDistanceSquared(worker.getPosition(), buildingPos));
 
+
                 //Then add the ladder position as the latest node.
-                proxyList.add(new BlockPos(ladderPos.getX(), level.getDepth(), ladderPos.getZ()));
+                proxyList.add(
+                  new BlockPos(
+                                ladderPos.getX() + building.getVectorX() * OTHER_SIDE_OF_SHAFT,
+                          level.getDepth(),
+                          ladderPos.getZ() + building.getVectorZ() * OTHER_SIDE_OF_SHAFT));
+
+                if(level.getRandomNode() != null && level.getRandomNode().getParent() != null)
+                {
+                    final List<BlockPos> nodesToTarget = new ArrayList<>();
+                    com.minecolonies.coremod.entity.ai.citizen.miner.Node currentNode = level.getNode(level.getRandomNode().getParent());
+                    while (new Point2D.Double(currentNode.getX(), currentNode.getZ()) != currentNode.getParent() && currentNode.getParent() != null)
+                    {
+                        nodesToTarget.add(new BlockPos(currentNode.getX(), levelDepth, currentNode.getZ()));
+                        currentNode = level.getNode(currentNode.getParent());
+                    }
+
+                    for (int i = nodesToTarget.size() - 1; i >= 0; i--)
+                    {
+                        proxyList.add(nodesToTarget.get(i));
+                    }
+                }
+
                 return newProxy;
             }
-            //If he is on the same Y level as his target and both underground, don't use a proxy. Just don't.
+            //If he is on the same Y level as his target and both underground.
             else if (targetY <= levelDepth)
             {
+                double closestNode = Double.MAX_VALUE;
+                Node lastNode = null;
+                for(final Map.Entry<Point2D, Node> node : level.getNodes().entrySet())
+                {
+                    final double distanceToNode = node.getKey().distance(worker.getPosition().getX(), worker.getPosition().getZ());
+                    if(distanceToNode < closestNode)
+                    {
+                        lastNode = node.getValue();
+                        closestNode = distanceToNode;
+                    }
+                }
+
+                if(lastNode != null && lastNode.getParent() != null)
+                {
+                    com.minecolonies.coremod.entity.ai.citizen.miner.Node currentNode = level.getNode(lastNode.getParent());
+                    while (new Point2D.Double(currentNode.getX(), currentNode.getZ()) != currentNode.getParent() && currentNode.getParent() != null)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Use the "equals" method if value comparison was intended. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1698)